### ISSUE TITLE
Update Offers Same File Every Time

### ISSUE THREAD

**Issue by fanuch** [twpayne/chezmoi#1357]
## What exactly are you trying to do?

Run `chezmoi update`  only with valid changes.
Update keeps showing me an empty "diff" on all systems for my vimrc file

## What have you tried so far?

`git pull; git status; git diff`

## Where else have you checked for solutions?

* [x] I have read [chezmoi's how-to guide](https://github.com/twpayne/chezmoi/blob/master/docs/HOWTO.md), and not found the answer.
* [x] I have read [chezmoi's FAQ](https://github.com/twpayne/chezmoi/blob/master/docs/FAQ.md), and not found the answer.
* [x] I have searched [chezmoi's reference guide](https://github.com/twpayne/chezmoi/blob/master/docs/REFERENCE.md), and not found the answer.
* [x] Other, please give details. Stack Overflow

## Output of any commands you've tried with `--verbose` flag

```console
$ chezmoi update --verbose
Already up to date.
Current branch master is up to date.
.vimrc has changed since chezmoi last wrote it [diff,overwrite,all-overwrite,skip,quit]? overwrite
# Rerun the same command instantly
$ chezmoi update --verbose
Already up to date.
Current branch master is up to date.
.vimrc has changed since chezmoi last wrote it [diff,overwrite,all-overwrite,skip,quit]? overwrite
```

## Output of `chezmoi doctor`

```console
$ chezmoi doctor
RESULT   CHECK                MESSAGE
ok       version              v2.0.7, commit 1ce19442a526747d53317a4cd5764f4e4a60a216, built at 2021-03-31T00:25:36Z, built by goreleaser
ok       os-arch              linux/amd64
ok       config-file          /home/ari/.config/chezmoi/chezmoi.toml is a file
ok       source-dir           /home/ari/.local/share/chezmoi is a directory
ok       suspicious-entries   no suspicious entries found in /home/ari/.local/share/chezmoi
ok       dest-dir             /home/ari is a directory
ok       shell                found /usr/bin/zsh
ok       editor               found /usr/bin/vim
ok       git-cli              found /usr/bin/git, version 2.25.1
ok       merge-cli            found /usr/bin/vimdiff
ok       age-cli              age not found in $PATH
ok       gnupg-cli            found /usr/bin/gpg, version 2.2.19
info     1password-cli        op not found in $PATH
info     bitwarden-cli        bw not found in $PATH
info     gopass-cli           gopass not found in $PATH
info     keepassxc-cli        keepassxc-cli not found in $PATH
info     keepassxc-db         not set
info     lastpass-cli         lpass not found in $PATH
info     pass-cli             pass not found in $PATH
ok       vault-cli            found /usr/bin/vault, version 1.8.1
info     secret-cli           not set
```

## Additional context

Git has no changes to it but for some reason, update will not stop prompting changes on vimrc:
```
➜ git diff
➜ echo $?
0
➜ chezmoi diff
➜ echo $?
0
➜ chezmoi update
Already up to date.
Current branch master is up to date.
.vimrc has changed since chezmoi last wrote it [diff,overwrite,all-overwrite,skip,quit]? overwrite
```

**Comment by twpayne** [comment]
Thanks for reporting this. I've seen the same thing myself a couple of times, but don't understand why it's happening yet. I think this is a bug in chezmoi.

**Comment by fanuch** [comment]
I was apprehensive about posting this thinking that I had done something ignorant. 
Thanks for confirming it's a bug!

Is there anything I could to do help resolve this?
Unfortunately, `--verbose` doesn't give away any hints

**Comment by twpayne** [comment]
#1380 adds a bunch of debug info to `chezmoi apply` which should help us understand the problem. The extra info is only shown with the `--debug` flag and is not part of a release yet. If you can build and run chezmoi from source then I'd be very happy to see the output of `chezmoi --debug ~/.vimrc`, otherwise I'll release the next version with the extra debug info in the next few days.

**Comment by fanuch** [comment]
Thanks for that debug addition.
For some reason, the issue is resolved on one of my machines, but still presents on another.

Here is the `chezmoi apply --debug ~/.vimrc` for the **working** and **broken** set ups!

The only discernible difference is this line in the broken version and not the working:
```
DBG Mkdir error="mkdir /home/ari/.config/chezmoi: file exists" name=/home/ari/.config/chezmoi perm=511
DBG Stat name=/home/ari/.config/chezmoi
```

[chezmoi_vimrc_debug_broken.txt](https://github.com/twpayne/chezmoi/files/7099625/chezmoi_vimrc_debug_broken.txt)
[chezmoi_vimrc_debug_working.txt](https://github.com/twpayne/chezmoi/files/7099626/chezmoi_vimrc_debug_working.txt)

**Comment by twpayne** [comment]
Thanks for running the command. If the binary included the change from #1380 then I would expect to see a line matching `defaultPreApplyFunc` in the logs, something like:

```
2021-09-02T18:29:18+02:00 DBG defaultPreApplyFunc actualEntryState={"ContentsSHA256":"735015e26ef0e6e77ea524dfa103f692d7a3f351db79c9c44b1d424c74b10639","Mode":420,"Type":"file","contents":"call plug#begin()\n\nPlug '/usr/local/opt/fzf'\nPlug 'AndrewRadev/s..."} lastWrittenEntryState={"ContentsSHA256":"735015e26ef0e6e77ea524dfa103f692d7a3f351db79c9c44b1d424c74b10639","Mode":420,"Type":"file"} targetEntryState={"ContentsSHA256":"735015e26ef0e6e77ea524dfa103f692d7a3f351db79c9c44b1d424c74b10639","Mode":420,"Type":"file","contents":"call plug#begin()\n\nPlug '/usr/local/opt/fzf'\nPlug 'AndrewRadev/s..."} targetRelPath=.vimrc
```

Are you sure that you've built chezmoi from the latest `master` branch? Alternatively you can download a binary built by chezmoi's CI system [here](https://github.com/twpayne/chezmoi/actions/runs/1192221979#artifacts).

That said, in the "working" log there's a call to update chezmoi's persistent state, which I think is due to [this logic that mitigates a bug in chezmoi prior to 2.0.10](https://github.com/twpayne/chezmoi/blob/master/internal/chezmoi/sourcestate.go#L479-L502). This would imply that the "working" chezmoi maybe isn't actually working, just hiding the problem.

**Comment by fanuch** [comment]
In my sleepiness to get this done I might not have successfully build and sourced that PR sorry.

> Is this also a possibility? i.e. that that same edits have been made to both the source and target entry states?

I do think that's likely. I feel like I might have triggered it after a `merge` or something

I'll try build and run again

**Comment by fanuch** [comment]
I'll be honest, I'm having a real hard time with running the PR from source.
[Instructions on site](https://www.chezmoi.io/docs/install/#from-source) only reference `@stable` and it isn't clear from my searches how to reference a commit, branch, or `pull/` (request).

Need to work on my `Go` op skills ...

So I've ran:

```
$ go install github.com/twpayne/chezmoi@latest
$ ~/golang/bin/chezmoi --version
chezmoi version dev
$ ~/golang/bin/chezmoi apply --verbose --debug ~/.vimrc

$
```

In `dev` version I cannot replicate the issue!

Also, I noticed releases has version 2.2.0 and my version is 2.0.7

Ran an `chezmoi upgrade` and I don't the error either :/ 

```
$ chezmoi upgrade
chezmoi: already at the latest version (2.2.0)

$ chezmoi update
Already up to date.
Current branch master is up to date.
```

I can't replicate this issue on `dev` or `latest` release so I will need to close this one off.
Thanks again for you help!
